File: kotlin/digitalwallet/src/main/kotlin/digitalwallet/services/WalletsService.kt
package digitalwallet.services

import digitalwallet.data.common.InvestmentRequest
import digitalwallet.data.common.LiquidationRequest
import digitalwallet.data.common.Logger
import digitalwallet.data.common.ProcessTransactionRequest
import digitalwallet.data.common.exceptions.TransactionFailed
import digitalwallet.data.enums.SubwalletType
import digitalwallet.data.enums.TransactionStatus
import digitalwallet.data.enums.TransactionType
import digitalwallet.data.enums.WalletType
import digitalwallet.repo.InvestmentPolicyRepo
import digitalwallet.repo.WalletFilter
import digitalwallet.repo.WalletsRepo

class InvestmentFailed(message: String) : Exception()
class LiquidationFailed(message: String) : Exception()

class WalletsService(
    private val walletsRepo: WalletsRepo,
    private val investmentPolicyRepo: InvestmentPolicyRepo,
    private val transactionsService: TransactionsService,
    private val investmentService: InvestmentService,
) {
    private val logger = Logger()

    suspend fun invest(request: InvestmentRequest) {
        val wallet = walletsRepo.find(
            WalletFilter(
                customerId = request.customerId,
                type = WalletType.REAL_MONEY
            )
        ).firstOrNull() ?: throw NoSuchElementException("Wallet not found for customer ${request.customerId}")

        val processTransactionRequest = ProcessTransactionRequest(
            amount = request.amount,
            idempotencyKey = request.idempotencyKey,
            originatorWalletId = wallet.id,
            originatorSubwalletType = SubwalletType.REAL_MONEY,
            type = TransactionType.HOLD,
        )

        val transaction = transactionsService.processTransaction(processTransactionRequest)

        if (transaction.status != TransactionStatus.PROCESSING) {
            throw InvestmentFailed("Failed to process investment ${request.idempotencyKey}")
        }
    }

    suspend fun liquidate(request: LiquidationRequest) {
        val wallet = walletsRepo.find(
            WalletFilter(
                customerId = request.customerId,
                type = WalletType.INVESTMENT
            )
        ) .firstOrNull() ?: throw NoSuchElementException("Wallet not found for customer ${request.customerId}")
        val investmentPolicy = investmentPolicyRepo.findById(wallet.policyId) ?: throw NoSuchElementException("Policy ${wallet.policyId} not found")

        try {
        investmentService.holdWithPolicy(
                InvestmentMovementRequest(
                amount = request.amount,
                idempotencyKey = request.idempotencyKey,
                walletId = wallet.id,
                investmentPolicy = investmentPolicy,
               )
            )
        } catch (e: TransactionFailed) {
            throw LiquidationFailed(e.message.toString())
        }
    }
}

File: scala/digitalwallet/app/src/main/scala/core/services/WalletsService.scala
package core.services

import core.domain.entities.{Transaction, Wallet}
import core.domain.enums.{BalanceType, SubwalletType, TransactionStatus, TransactionType, WalletType}
import core.domain.model.{InvestmentRequest, LedgerQuery, ProcessTransactionRequest}
import core.errors.{InvestmentFailedError, WalletNotFound}
import ports.{WalletFilter, WalletsDatabase}

class WalletsService(walletsRepo: WalletsDatabase, ledgerService: LedgerService, transactionsService: TransactionsService) {
  def getAvailableBalance(wallet: Wallet): BigDecimal = {
    val ledgerQuery = wallet.walletType match {
      case WalletType.RealMoney =>
        List(
          LedgerQuery(
            subwalletType = SubwalletType.RealMoney,
            balanceType = BalanceType.Available
          )
        )
      case WalletType.Investment =>
        List(
          LedgerQuery(
            subwalletType = SubwalletType.Bonds,
            balanceType = BalanceType.Available,
          ),
          LedgerQuery(
            subwalletType = SubwalletType.Stock,
            balanceType = BalanceType.Available,
          ),
          LedgerQuery(
            subwalletType = SubwalletType.RealEstate,
            balanceType = BalanceType.Available,
          ),
          LedgerQuery(
            subwalletType = SubwalletType.Cryptocurrency,
            balanceType = BalanceType.Available,
          )
        )
      case WalletType.EmergencyFunds =>
        List(
          LedgerQuery(
            subwalletType = SubwalletType.EmergencyFunds,
            balanceType = BalanceType.Available
          )
        )
    }

    ledgerService.getBalance(wallet.id, ledgerQuery)
  }

  def invest(request: InvestmentRequest): Either[InvestmentFailedError, Transaction] = {
    val wallets = walletsRepo.find(WalletFilter(customerId = Some(request.customerId), walletType = Some(WalletType.RealMoney)))

    wallets match {
      case _ => Left(WalletNotFound(s"Wallet not found for customer ${request.customerId}"))
      case List(wallet) =>
        val createTransactionRequest = ProcessTransactionRequest(
          amount = request.amount,
          idempotencyKey = request.idempotencyKey,
          originatorWalletId = wallet.id,
          originatorSubwalletType = SubwalletType.RealMoney,
          transactionType = TransactionType.Hold
        )

        val transaction = transactionsService.create(createTransactionRequest)

        for {
          processTransactionTuple <- transactionsService.process(transaction).left.map { e =>
            transactionsService.updateStatus(transaction.id, TransactionStatus.Failed)
            InvestmentFailedError(e.message)
          }
          executedTransaction <- transactionsService.execute(processTransactionTuple).left.map(e => InvestmentFailedError(e.message))
        } yield executedTransaction
    }
  }
}
